70_FLIPFluids/app_resources/compute/diffusion.comp.hlsl

```diff
@@ -0,0 +1,142 @@
+[[vk::binding(b_dDiffOutBuffer, s_d)]] RWStructuredBuffer<float4> gridDiffusionOutBuffer;
+
+[numthreads(WorkgroupSize, 1, 1)]
+void setAxisCellMaterial(uint32_t3 ID : SV_DispatchThreadID)
+{
+    uint cid = ID.x;
+    int3 cellIdx = flatIdxToCellIdx(cid, gridData.gridSize);
+
+    uint cellMaterial = cellMaterialBuffer[cid];
+
+    uint this_cm = getCellMaterial(cellMaterial);
+    uint xp_cm = getXPrevMaterial(cellMaterial);
+    uint yp_cm = getYPrevMaterial(cellMaterial);
+    uint zp_cm = getZPrevMaterial(cellMaterial);
+
+    uint3 cellAxisType;
+    cellAxisType.x = 
+        isSolidCell(this_cm) || isSolidCell(xp_cm) ? CM_SOLID :
+        isFluidCell(this_cm) || isFluidCell(xp_cm) ? CM_FLUID :
+        CM_AIR;
+    cellAxisType.y = 
+        isSolidCell(this_cm) || isSolidCell(yp_cm) ? CM_SOLID :
+        isFluidCell(this_cm) || isFluidCell(yp_cm) ? CM_FLUID :
+        CM_AIR;
+    cellAxisType.z = 
+        isSolidCell(this_cm) || isSolidCell(zp_cm) ? CM_SOLID :
+        isFluidCell(this_cm) || isFluidCell(zp_cm) ? CM_FLUID :
+        CM_AIR;
+
+    uint3 cmAxisTypes = 0;
+    setCellMaterial(cmAxisTypes, cellAxisType);
+
+    axisCellMaterialOutBuffer[cid].xyz = cmAxisTypes;
+}
+
+[numthreads(WorkgroupSize, 1, 1)]
+void setNeighborAxisCellMaterial(uint32_t3 ID : SV_DispatchThreadID)
+{
+    uint cid = ID.x;
+    int3 cellIdx = flatIdxToCellIdx(cid, gridData.gridSize);
+
+    uint3 axisCm = (uint3)0;
+    uint3 this_axiscm = getCellMaterial(axisCellMaterialInBuffer[cid].xyz);
+    setCellMaterial(axisCm, this_axiscm);
+
+    uint3 xp_axiscm = cellIdx.x == 0 ? (uint3)CM_SOLID : getCellMaterial(axisCellMaterialInBuffer[cellIdxToFlatIdx(cellIdx + int3(-1, 0, 0), gridData.gridSize)].xyz);
+    setXPrevMaterial(axisCm, xp_axiscm);
+
+    uint3 xn_axiscm = cellIdx.x == gridData.gridSize.x - 1 ? (uint3)CM_SOLID : getCellMaterial(axisCellMaterialInBuffer[cellIdxToFlatIdx(cellIdx + int3(1, 0, 0), gridData.gridSize)].xyz);
+    setXNextMaterial(axisCm, xn_axiscm);
+
+    uint3 yp_axiscm = cellIdx.y == 0 ? (uint3)CM_SOLID : getCellMaterial(axisCellMaterialInBuffer[cellIdxToFlatIdx(cellIdx + int3(0, -1, 0), gridData.gridSize)].xyz);
+    setYPrevMaterial(axisCm, yp_axiscm);
+
+    uint3 yn_axiscm = cellIdx.y == gridData.gridSize.y - 1 ? (uint3)CM_SOLID : getCellMaterial(axisCellMaterialInBuffer[cellIdxToFlatIdx(cellIdx + int3(0, 1, 0), gridData.gridSize)].xyz);
+    setYNextMaterial(axisCm, yn_axiscm);
+
+    uint3 zp_axiscm = cellIdx.z == 0 ? (uint3)CM_SOLID : getCellMaterial(axisCellMaterialInBuffer[cellIdxToFlatIdx(cellIdx + int3(0, 0, -1), gridData.gridSize)].xyz);
+    setZPrevMaterial(axisCm, zp_axiscm);
+
+    uint3 zn_axiscm = cellIdx.z == gridData.gridSize.z - 1 ? (uint3)CM_SOLID : getCellMaterial(axisCellMaterialInBuffer[cellIdxToFlatIdx(cellIdx + int3(0, 0, 1), gridData.gridSize)].xyz);
+    setZNextMaterial(axisCm, zn_axiscm);
+
+    axisCellMaterialOutBuffer[cid].xyz = axisCm;
+}
```

Comment:
this could be kernel fused, if you used 8x8x8 workgroup, preloaded 10x10x10 (1 cell skirt) updated the axis cell in smem and then did the neighbor axis update on the inner 8x8x8 of the 10x10x10

See https://youtu.be/Ol_sHFVXvC0?si=RrP91pPpMzODMEDi&t=1585